Support for new data types when reading/writing networks

KasiaKoz opened the pull request:
Adds support (read/write) for:
- CSV (network nodes and links csv tables and GTFS-like tables for the PT schedule)
  - Comprehensive export for network nodes and links
  - Some data loss (relations of stops and routing on the network) for schedule
  - read csv for schedule is equivalent to reading GTFS, it is a new method that is more responsive to the data stored and forgiving for data missing in GTFS (the list of reindexing warnings when reading it to schedule object will disappear @Arupwkeu )
- JSON 
  - The most comprehensive export for network and schedule
- GeoJSON
  -  The indexing is now consistent with link indexing for network
  - read only for network, format not very supportive for all of the schedule data
  - write for geojsons is the same as before, but can now be done via a class method `network.write_to_geojson(path)`

Adds nice class methods for both `Network` and `Schedule` objects to put them in the formats mentioned above
- `to_json`
- `to_geodataframe`

Changes MATSim read methods - **breaking change**. All read methods are now stored in the read module, which can be imported in the following way, e.g
```python
from genet import read_matsim
```
where you have a selection of read methods, all return a genet object from that data type. The syntax to read a MATSim network changes from:
```python
from genet import Network

n = Network('epsg:27700')
n.read_matsim_network(path_to_matsim_network)
n.read_matsim_schedule(path_to_matsim_schedule, path_to_matsim_vehicles)
```
to
```python
from genet import read_matsim

n = read_matsim(
    path_to_network=path_to_matsim_network, 
    epsg='epsg:27700', 
    path_to_schedule=path_to_matsim_schedule, 
    path_to_vehicles=path_to_matsim_vehicles
)
```
OSM reading method also moved to the `read` module for consistence.

Adds a lot of new test/example data and jupyter notebooks which which show examples and talk about the schema if you want to read data and limitations of what gets saved if you're saving it out.

mfitz reviewed (approved):
Looks good.

I've made a comment or two in the tests asking if there are any ways we could be clearer about why we have some of the expectations we're asserting against, which is a question that applied to a fair number of different tests.

I also noticed we have an XFAIL test (`tests/test_core_schedule_elements.py::test_splitting_service_edge_case_on_direction_results_in_two_directions XFAIL`) and a *lot* of warnings:

```
669 passed, 1 xfailed, 440 warnings in 550.78s (0:09:10)
```

Might be worth cleaning those up in a separate story.

KasiaKoz commented on genet/inputs_handler/read.py:
```diff
@@ -0,0 +1,330 @@
+        json_data = json.load(json_file)
+    for node, data in json_data['nodes'].items():
+        try:
+            del data['geometry']
```
Depending on where this json comes from, all of the geometry might be missing (it's not a requirement for a json network format) or none of it will be missing (genet exports node geometry to json, so if you read it back, it'll all be there)

KasiaKoz commented on genet/outputs_handler/csv.py:
```diff
@@ -0,0 +1,10 @@
+        expected schema
+    :return: genet.Network object
+    """
+    pass
```
lol, I put a placeholder there and then ended up putting the method somewhere else 😂

KasiaKoz commented on notebooks/4.3. Using Network - Routing.ipynb:
```diff
@@ -222,7 +231,7 @@
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.8.6"
+   "version": "3.7.0"
```
probably not deliberate, I had to make a new virtual environment for genet at one point. I think this is better (genet's requirement is >=3.7 so I should probably use the lowest version ?)

mfitz replied:
I don't think it really matters which version you use, so long as you're deliberate and consistent.

KasiaKoz commented:
Thanks @mfitz, I fixed up a few things and added a card [LAB-1179](https://arupdigital.atlassian.net/browse/LAB-1179) for a more serious maintenance of genet tests. The XFAIL is there to stay for now though, it's an edge case marker for method that does most of the job, but there is no requirement to cover that edge case right now.
